### ISSUE TITLE
`circle (40, "solid", "red")` generates lousy feedback -- complains about the first comma

### ISSUE THREAD

**Issue by blerner** [brownplt/pyret-lang#1113]
This isn't new; we've complained about this before.  And "unfortunately", `(a, b, c)` isn't tuple syntax, so this doesn't parse as `id-expr tuple-expr`, so Jack's reparsing infrastructure can't quite handle it.  I'm not sure what technically we can do better on this, right now.

**Comment by jswrenn** [comment]
> And "unfortunately", (a, b, c) isn't tuple syntax

Couldn't we just add a `malformed-tuple` to the grammar,
```
malformed-tuple: (PARENSPACE|PARENAFTERBRACE) comma-binops RPAREN
```
and a corresponding variant to `Expr`, and report a specialized variant of `same-line` when the second expression is `malformed-tuple`? 

If we had this, could also introduce a well-formedness error targeted at people who are used to a parentheses-based tuple syntax to inform them that pyret uses curly braces (à la @shriram's recommendation in #1112 regarding exponentiation).

**Comment by blerner** [comment]
No, that makes the grammar ambiguous.

**Comment by jswrenn** [comment]
I recall some discussion about that but couldn't find anything on slack. What's the ambiguity?

**Comment by blerner** [comment]
`fun-expr` parses the same as `id-expr malformed-tuple`

**Comment by jswrenn** [comment]
Are you sure? I think there are two features that keep that case unambiguous. First, it is a syntax error for the name in a `fun-expr` to be followed by a space; e.g.:
```
fun foo (a):
  1
end
```
Second, `fun-expr` mandates the presence of the keyword `fun`.

**Comment by blerner** [comment]
Sorry, `app-expr`, not `fun-expr`

**Comment by jswrenn** [comment]
I'm not sure about that, either. An `app-expr` insists that the first `expr` is _not_ followed by a space, but for the error we're discussing here, the first `expr` is _always_ followed by a space.

If the grammar is defined like so (`app-expr` is what we have, `tuplish` is the proposed new rule),
```
app-expr : expr PARENNOSPACE opt-comma-binops RPAREN
tupleish : expr PARENSPACE   comma-binops     RPAREN
```
then aren't these cases clearly disambiguated the presence or absence of a space?

We play this trick elsewhere in the grammar; e.g., the definition of the `paren-expr` rule:
```
# paren-exprs must be preceded by a space, so as not be be confused with
# function application
paren-expr: (PARENSPACE|PARENAFTERBRACE) binop-expr RPAREN
```

**Comment by blerner** [comment]
Yes, I'm well aware we play that game elsewhere in the grammar ... and it's a brittle hack that I don't trust and never have.  If you introduce this parenspace/parennospace hack for "tupleish" expressions, then you risk this ambiguity for function calls, data variant definitions, and _you make the grammar really brittle_ to any future refactoring.  I *really, really* don't want to introduce this kind of hack any further.

**Comment by jswrenn** [comment]
I'm not proposing that we make `tupleish` _the_ syntax for tuples; I'm proposing that we add it to the grammar _solely_ for the purpose of displaying a better error message. 

Since we have never considered changes to error message text a language-breaking change, it should not a big deal if we need to drop it at some time in the future. At that point, we'll just be back to where we are now. Until then, we could be giving a far better error report for this scenario.

I'm not trying to be obstinate, but I'm compelled to advocate for better error reports if there aren't hard technical obstacles to doing so. This bug report seems like it has a really simple fix.

**Comment by blerner** [comment]
On the contrary.  Adding something to the grammar solely for error messages is *not* a simple fix, because I'm not 100% confident that it won't inadvertently have any knock-on effects on the grammar, or cause something else to become ambiguous.  Otherwise we wouldn't have needed to restrict this case in the first place!

I'm *not* willing to do that.  For now, I'm stating flatly that the grammar can't be augmented just to handle error cases.  We need to find another technical solution; this one isn't it.

**Comment by shriram** [comment]
I'm fully with Ben here. The danger of any hacks that we have not thought
through 200% is that something else will break even more catastrophically.

**Comment by jswrenn** [comment]
@shriram, I absolutely agree that we _should_ think these things through 200%, which is why I've been very concrete in this discussion. However, @blerner's last comment isn't one about a technical obstacle, it's a statement of policy that shuts down a technical discussion:
> For now, I'm stating flatly that the grammar can't be augmented just to handle error cases. We need to find another technical solution; this one isn't it.

If that really is our policy, then whether or not we've "thought [the technical aspects] through 200%" is irrelevant.

**Comment by blerner** [comment]
Yes, I'm shutting this discussion down, *because* we clearly had thought this through at some point in the past -- that's why there's a distinction between PARENSPACE and PARENNOSPACE.  And the fact that we can't remember exactly what the knock-on effects are, why the distinction remains, and whether it'd screw up the grammar elsewhere, means that we're *not* 200% sure about the effects, so this is *not* an acceptable solution.

**Comment by shriram** [comment]
Ben is stating a principle: that the grammar should remain the pure grammar
(as much as possible), not grow various warts to enable certain kinds of
error reporting. That is, he wants a separation of concerns. Are you
disagreeing with that?

But he's also going farther than that. He says _he_, the grammar's owner
and maintainer, can't think through all the cases to be 200% sure. Are you
disagreeing with him? Do you think he has thought through all the cases and
is actually 200% sure, without realizing it?

**Comment by jswrenn** [comment]
> Ben is stating a principle: that the grammar should remain the pure grammar (as much as possible), not grow various warts to enable certain kinds of error reporting. That is, he wants a separation of concerns. Are you disagreeing with that?

Emphatically: _yes_. In order to give good syntax error reports, we _have_ repeatedly expanded the grammar to accept programs that are not what we want to be syntactically valid, and then immediately curtailed those programs via well-formedness checks. For example:
- The grammar allows empty blocks, but they are immediately forbidden with a well-formedness check.
- The grammar allows two expressions on the same line, but they are immediately forbidden with a well-formedness check.
- The grammar allows adding a `check` block to a `lam`, but they are are immediately forbidden with a well-formedness check.
- The grammar allows defining methods that take no parameters, but they are immediately forbidden with a well-formedness check.

The more constrained our grammar is, the _more_ tied to error reporting it becomes, because we're at the mercy of whatever the generated parser throws up.

(This is rehashing a discussion that happened on Slack last week and with very little consensus. That last line paraphrases [a comment](https://pyret.slack.com/archives/C1PK2JDV2/p1501098580971175) by @jpolitz, who later also notes that we need to very carefully consider the danger of ambiguities.)

> But he's also going farther than that. He says _he_, the grammar's owner and maintainer, can't think through all the cases to be 200% sure. Are you disagreeing with him? Do you think he has thought through all the cases and is actually 200% sure, without realizing it?

Of course not. But I don't view "we can't remember exactly what the knock-on effects are" as an invitation to "shutting this discussion down"; I view it as a pretty alarming cause for _more_ discussion. If Pyret's grammar has "brittle hacks" we should document why they are brittle.

**Comment by blerner** [comment]
Those error cases, as was mentioned in that Slack discussion, are *not* syntax errors: they're *opinionated complaints* by the Pyret development folks that those programs are in poor taste.  They don't invite parsing ambiguity, they don't compromise the grammar, and they *happen* to produce friendlier errors.  (You've also got causality wrong: we allowed most of those problems as valid programs, before they were prohibited by well-formedness.)

Grammars are *hard*, ambiguity detection is (impossibly) *hard*, and I am *not* confident that we can deduce what the consequences of these changes would be.  I'm not even 100% confident that what we have right now is 100% correct (as in, does what we as experienced developers intuitively would want), but I am not willing to revise anything nearby this part of the grammar.  "If it ain't broke, don't fix it"

**Comment by jswrenn** [comment]
> Those error cases, as was mentioned in that Slack discussion, are not syntax errors: they're opinionated complaints by the Pyret development folks that those programs are in poor taste.

You did argue that, but that was also the point of greatest contention. A method defined without parameters isn't an "opinionated complaint", it's semantically malformed; it is literally impossible to call a method without passing at least one argument to it. So we had to forbid it, and we had a choice:
 - forbid it with the grammar, or
 - forbid it with a well-formedness check.

We chose to forbid it with a well-formedness check.

Additionally, @jpolitz raised a different case of a different well-formedness error that he saw as a foray into "ungrammar" territory.

I just don't think we should adopt a policy as sweeping as "the grammar can't be augmented just to handle error cases" without considerable scrutiny, especially when there's some disagreement about whether we _already_ do this. 

-----------------

> "If it ain't broke, don't fix it"

Agreed, but this:
> we can't remember exactly what the knock-on effects are, why the distinction remains, and whether it'd screw up the grammar elsewhere, means that we're not 200% sure about the effects

doesn't sound "fixed". We _should_ know these things. We won't ever be "200% sure about the effects" if we close this subject off to discussion.

---------

I'm all for alternative technical solutions, but since this bug report documents a failure that happens at the parsing stage, the alternative technical solution almost certainly must be at the parsing stage. We have only two things to manipulate at this stage:
 1. the grammar
 2. the mechanism by which parsing occurs

If we're going to rule out all changes to the former, we should move this discussion of this this particular bug to changes to the latter.
